### PR TITLE
refactor: Foundation

### PR DESCRIPTION
**Pull request by MartinZikmund**
GitHub Issue (If applicable): closes #

## PR Type

What kind of change does this PR introduce?

## What is the current behavior?

Some parts of `Windows.Foundation` are not implemented.

## What is the new behavior?

Everything in `Windows.Foundation` is implemented!

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.Foundation/Collections/Internal/NullableKeyDictionary.cs [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+	public bool TryGetValue(TKey? key, out TValue value)
+	{
+		if (key != null)
+		{
+			return _dictionary.TryGetValue(key, out value!);
+		}
+		else
+		{
+			value = _containsNullKey ? _nullKeyValue : default!;
```
Comment (Youssef1313) [inline src/Uno.Foundation/Collections/Internal/NullableKeyDictionary.cs:72]: I guess that should work?

```suggestion
			value = _nullKeyValue;
```

File: src/Uno.Foundation/Collections/Internal/NullableKeyDictionary.cs [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	public void CopyTo(KeyValuePair<TKey, TValue>[] array, int arrayIndex)
+	{
+		foreach (var pair in this)
+		{
+			array[arrayIndex] = pair;
+			arrayIndex++;
+		}
+	}
+
+	public bool Remove(KeyValuePair<TKey, TValue> item) => Contains(item) ? Remove(item.Key) : false;
```
Comment (Youssef1313) [inline src/Uno.Foundation/Collections/Internal/NullableKeyDictionary.cs:179]:
```suggestion
	public bool Remove(KeyValuePair<TKey, TValue> item) => Contains(item) && Remove(item.Key);
```

File: src/Uno.Foundation/Extensions/UriExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+using System.Linq;
+using System.Text;
+
+namespace Uno.Extensions;
+
+public static class UriExtensions
+{
+	private const int EscapeDataStringCharactersMaxLength = 10000;
+
+	public static IDictionary<string, string> GetParameters(this Uri uri)
```
[the remark this answers is not in the document]
Comment (MartinZikmund) [reply src/Uno.Foundation/Extensions/UriExtensions.cs:9]: Done!

File: src/Uno.Foundation/Collections/ValueSet.cs [hunk 4]
```diff
@@ -1,34 +1,71 @@
...
+/// Implements a map with keys of type String and values of type Object.
+/// </summary>
+public sealed partial class ValueSet :
+	IDictionary<string, object>,
+	IEnumerable<KeyValuePair<string, object>>,
+	IObservableMap<string, object>,
+	IPropertySet
 {
-	private readonly Dictionary<string, object?> _dictionary = new();
+	private readonly Dictionary<string, object> _properties = new Dictionary<string, object>();
```
Comment (jeromelaban) [inline src/Uno.Foundation/Collections/ValueSet.cs:14]: Why did you remove the nullability on this one?

File: src/Uno.Foundation/Metadata/ApiInformation.cs [hunk 5]
```diff
@@ -1,164 +1,164 @@
 #nullable enable
 
-using System;
-using System.Collections.Generic;
-using System.Linq;
 using System.Reflection;
 using Uno.Foundation.Logging;
-
-namespace Windows.Foundation.Metadata
+//TODO
```
Comment (jeromelaban) [inline src/Uno.Foundation/Metadata/ApiInformation.cs:5]: What's the todo for?

File: src/Uno.Foundation/Metadata/AttributeNameAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+/// Indicates the name of the attribute.
+/// </summary>
+[AttributeUsage(AttributeTargets.Class)]
+public partial class AttributeNameAttribute : Attribute
+{
+	/// <summary>
+	/// Creates and initializes a new instance of the attribute.
+	/// </summary>
+	/// <param name="A_0">One or more of the enumeration values.</param>
+	public AttributeNameAttribute(string A_0) : base()
```
Comment (jeromelaban) [inline src/Uno.Foundation/Metadata/AttributeNameAttribute.cs:13]: A_0 ?

Comment (Youssef1313) [reply src/Uno.Foundation/Metadata/AttributeNameAttribute.cs:13]: @jeromelaban Weird naming, but that matches UWP per documentation https://learn.microsoft.com/en-us/uwp/api/windows.foundation.metadata.attributenameattribute.-ctor?view=winrt-22621

Comment (MartinZikmund) [reply src/Uno.Foundation/Metadata/AttributeNameAttribute.cs:13]: 🤷‍♂️🤷‍♂️😂

File: src/Uno.Foundation/Metadata/ApiInformation.cs (hunk 5 above)
Comment (MartinZikmund) [reply src/Uno.Foundation/Metadata/ApiInformation.cs:5]: Accidental leftover

File: src/Uno.Foundation/Collections/ValueSet.cs (hunk 4 above)
Comment (MartinZikmund) [reply src/Uno.Foundation/Collections/ValueSet.cs]: Unintentional, fixed!

**Comment by Youssef1313** [comment]
Still failing even after updating C# version in generator :/
I'm going to take a look @MartinZikmund .

**Comment by Youssef1313** [comment]
Looks like the project is still seen with LangVersion = 8

image

**Comment by Youssef1313** [comment]
I pushed a small update to your branch @MartinZikmund. Hope you're okay with that :)

**Comment by MartinZikmund** [comment]
Thanks @Youssef1313 , it seems much better! A few more fails remain, would you be able to take one more look please? I am unfortunately low on time this week, but would love to have this merged finally 😅

**Comment by Youssef1313** [comment]
@MartinZikmund When comparing the indexer type from UWP to the one implemented in Uno, we were comparing `object` vs `object?`, so they compared unequal, and then the generator thinks that the indexer is not implemented so it re-generated it. I adjusted the comparison in multiple places for nullability.

File: src/Uno.Foundation/Collections/PropertySet.cs [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+		}
+		return _dictionary.TryGetValue(key, out value);
+	}
+
+	/// <summary>
+	/// Gets or sets a value for the specified key.
+	/// </summary>
+	/// <param name="key">The key.</param>
+	/// <returns>Value.</returns>
+	public object? this[string key]
```
Comment (Youssef1313) [inline src/Uno.Foundation/Collections/PropertySet.cs:99]: This little question mark here was the reason for the generator failure :)

**Comment by MartinZikmund** [comment]
@Youssef1313 Thanks sir!

File: src/Uno.Foundation/IPropertyValue.cs [hunk 8]
```diff
@@ -1,133 +1,245 @@
...
+	/// Returns the date and time value stored as a property value.
+	/// </summary>
+	/// <returns>The value.</returns>
+	DateTimeOffset GetDateTime();
+
+	/// <summary>
+	/// Returns the array of date and time values stored as a property value.
+	/// </summary>
+	/// <param name="value">The array of values.</param>
+	void GetDateTimeArray(out DateTime[] value);
```
Comment (Youssef1313) [inline src/Uno.Foundation/IPropertyValue.cs:58]: Should the parameter here be `DateTimeOffset[]` instead of `DateTime[]`?

Comment (MartinZikmund) [reply src/Uno.Foundation/IPropertyValue.cs]: Indeed, there was a bug in UWP/WinUI docs for this! Fixed the code

**Review by Youssef1313 — COMMENTED** [review]
Quite large to review in depth, but overall LGTM.

Will let @jeromelaban approve.

**Comment by Youssef1313** [comment]
@MartinZikmund Can you fix the small conflict here?

@jeromelaban Is this ready to merge?

**Review by jeromelaban — APPROVED** [review]
